mm2src/lp_swap/trade_preimage.rs

```diff
@@ -297,6 +321,63 @@ impl TradePreimageRpcError {
         // so avoid unnecessary boilerplate code.
         TradePreimageRpcError::from(CheckBalanceError::from_trade_preimage_error(trade_preimage_err, ticker))
     }
+
+    pub fn from_maker_order_build_error(
+        maker_order_err: MakerOrderBuildError,
+        base: &str,
+        rel: &str,
+    ) -> TradePreimageRpcError {
+        match maker_order_err {
+            MakerOrderBuildError::BaseEqualRel => TradePreimageRpcError::BaseEqualRel,
+            MakerOrderBuildError::MaxBaseVolTooLow { actual, threshold } => TradePreimageRpcError::VolumeTooLow {
+                coin: base.to_owned(),
+                volume: actual.to_decimal(),
+                threshold: threshold.to_decimal(),
+            },
+            MakerOrderBuildError::PriceTooLow { actual, threshold } => TradePreimageRpcError::PriceTooLow {
+                price: actual.to_decimal(),
+                threshold: threshold.to_decimal(),
+            },
+            MakerOrderBuildError::RelVolTooLow { actual, threshold } => TradePreimageRpcError::VolumeTooLow {
+                coin: rel.to_owned(),
+                volume: actual.to_decimal(),
+                threshold: threshold.to_decimal(),
+            },
+            // The errors below may occur due to invalid dummy params.
+            error @ MakerOrderBuildError::MinBaseVolTooLow { .. }
+            | error @ MakerOrderBuildError::ConfSettingsNotSet
+            | error @ MakerOrderBuildError::MaxBaseVolBelowMinBaseVol { .. } => {
+                TradePreimageRpcError::InternalError(format!("Unexpected MakerOrderBuildError: {}", error))
+            },
+        }
+    }
+
+    pub fn from_taker_order_build_error(
+        taker_order_err: TakerOrderBuildError,
+        base: &str,
+        rel: &str,
+    ) -> TradePreimageRpcError {
+        match taker_order_err {
+            TakerOrderBuildError::BaseEqualRel => TradePreimageRpcError::BaseEqualRel,
+            TakerOrderBuildError::BaseAmountTooLow { actual, threshold } => TradePreimageRpcError::VolumeTooLow {
+                coin: base.to_owned(),
+                volume: actual.to_decimal(),
+                threshold: threshold.to_decimal(),
+            },
+            TakerOrderBuildError::RelAmountTooLow { actual, threshold } => TradePreimageRpcError::VolumeTooLow {
+                coin: rel.to_owned(),
+                volume: actual.to_decimal(),
+                threshold: threshold.to_decimal(),
+            },
+            // The errors below may occur due to invalid dummy params.
+            error @ TakerOrderBuildError::MinVolumeTooLow { .. }
```

Comment:
It will be a bit shorter to have
```rust
_ => TradePreimageRpcError::InternalError(format!("Unexpected TakerOrderBuildError: {}", taker_order_err))
```
The error message will be longer in this case, though. It's up to you whether you prefer to change it :slightly_smiling_face: 

PS: glad to know that `@` exists - was not aware of such thing :slightly_smiling_face:

Reply:
Yeah, good point, thanks! :slightly_smiling_face: